### PR TITLE
Add raw_coprocessor

### PR DESCRIPTION
**Pull request by andylokandy**
Issue: https://github.com/tikv/tikv/issues/9747

Blocking on https://github.com/pingcap/kvproto/pull/765

Add support for sending raw coprocessor requests.

## Details

- Rename `PreserveKey` to `PreserveShard` and remove `HasKeys`.
- Rename `CollectAndMatchKeys` to `CollectWithShard`
- Rename `ResponseAndKeys` to `ResponseWithShard`
- Add RawCoproceesorRequest wrapper for dynamically building custom request data after sharding

### CODE REVIEW

**Comment by andylokandy** [comment]
cc @Jibbow

File: src/raw/client.rs [hunk 1]
```diff
@@ -491,6 +492,28 @@ impl Client {
         plan.execute().await
     }
 
+    pub async fn coprocessor(
+        &self,
+        copr_name: String,
+        copr_version_req: String,
```
Comment (Jibbow) [inline src/raw/client.rs:498]: Should we maybe use `semver::VersionReq` here so that we can be sure that the user passes a valid string?

**Review by Jibbow — APPROVED** [review]

File: src/raw/client.rs (hunk 1 above)
Comment (andylokandy) [reply src/raw/client.rs:498]: Good idea!

File: src/transaction/requests.rs [hunk 2]
```diff
@@ -346,61 +342,62 @@ impl Shardable for kvrpcpb::PessimisticLockRequest {
...
+    ) -> Result<Self::Out> {
         input
             .into_iter()
-            .flat_map_ok(|mut resp| {
+            .flat_map_ok(|ResponseWithShard(mut resp, mutations)| {
                 let values = resp.take_values();
+                let values_len = values.len();
                 let not_founds = resp.take_not_founds();
-                let v: Vec<_> = if not_founds.is_empty() {
+                let kvparis = mutations
```
Comment (ekexium) [inline src/transaction/requests.rs:380]:
```suggestion
                let kvpairs = mutations
```

File: src/request/shard.rs [hunk 3]
```diff
@@ -30,21 +36,27 @@ pub trait Shardable {
     type Shard: Send;
 
     fn shards(
```
Comment (ekexium) [inline src/request/shard.rs:38]: I think there can be a better name like `group_shards_by_store`

File: src/request/shard.rs [hunk 4]
```diff
@@ -30,21 +36,27 @@ pub trait Shardable {
...
-        &self,
+        shard: Self::Shard,
         pd_client: &Arc<impl PdClient>,
     ) -> BoxStream<'static, Result<(Self::Shard, Store)>>;
 
+    fn get_shard(&self) -> Self::Shard;
```
Comment (ekexium) [inline src/request/shard.rs:43]: I think `get_` had better return a reference. Let the user clone it if necessary.

File: src/request/plan.rs [hunk 5]
```diff
@@ -315,102 +301,64 @@ where
...
+        Ok(ResponseWithShard(res, shard))
     }
 }
 
-pub trait HasKeys {
-    fn get_keys(&self) -> Vec<Key>;
-}
-
 // contains a response and the corresponding keys
 // currently only used for matching keys and values in pessimistic lock requests
```
Comment (ekexium) [inline src/request/plan.rs:336]: The comment needs update. And I think we can add a hint saying that the order of the elements in the response and in the shard must match.

**Review by ekexium — COMMENTED** [review]
I think an alternative of the method is to add a new plan that do some pre-pocessing. The plan can be placed between `MultiRegion` and `Dispatch`, so that we don't have to create a special request type.

We should merge #292 first.

File: src/raw/requests.rs [hunk 6]
```diff
@@ -316,6 +330,107 @@ impl Process<kvrpcpb::RawCasResponse> for DefaultProcessor {
...
+    ) -> Result<Self::Out> {
+        input?
+            .into_iter()
+            .map(|shard_resp| {
+                shard_resp.map(|ResponseWithShard(mut resp, ranges)| {
+                    (
+                        resp.take_data(),
+                        ranges
+                            .into_iter()
+                            .map(|range| range.start_key.into()..range.end_key.into())
```
[the remark this answers is not in the document]
Comment (andylokandy) [reply src/raw/requests.rs:425]: Empty end key is handled in TiKV. Besides, this is how `KeyRange` is defined in kvproto.

File: src/request/shard.rs (hunk 4 above)
Comment (andylokandy) [reply src/request/shard.rs:43]: If the shard is computed, `self` doesn't have the ownership, thus not able to return a reference. Anyway, the only usage of get_shard are `MultiRegion` and `PreserveShard` which will always clone it.

File: src/request/shard.rs (hunk 3 above)
Comment (andylokandy) [reply src/request/shard.rs:38]: I'd like to make a following PR to refactor these parts.

File: src/request/shard.rs (hunk 4 above)
Comment (ekexium) [reply src/request/shard.rs:43]: Then we can name it after something like `get_cloned_shard`

Comment (andylokandy) [reply src/request/shard.rs]: I reverted get_shard.

**Comment by andylokandy** [comment]
@ekexium Conflict resovled, PTAL

**Comment by andylokandy** [comment]
oh, stupid me...

**Comment by andylokandy** [comment]
> oh, stupid me...

Fixed in https://github.com/tikv/client-rust/commit/eb1d2da05cc857818e078bc3d91ce50343031642

**Pull request merged**